LicenseHeaderManager/PackageCommands/AddLicenseHeaderToAllFilesInSolutionCommand.cs

```diff
@@ -0,0 +1,100 @@
+
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using EnvDTE;
+using LicenseHeaderManager.Headers;
+using LicenseHeaderManager.Interfaces;
+using LicenseHeaderManager.SolutionUpdateViewModels;
+using LicenseHeaderManager.Utils;
+
+namespace LicenseHeaderManager.PackageCommands
+{
+  public class AddLicenseHeaderToAllFilesInSolutionCommand : ISolutionLevelCommand
+  {
+    private const string c_commandName = "Add LicenseHeader to all files in Solution";
+
+    private readonly LicenseHeaderReplacer _licenseReplacer;
+    private readonly SolutionUpdateViewModel _solutionUpdateViewModel;
+
+    public AddLicenseHeaderToAllFilesInSolutionCommand(LicenseHeaderReplacer licenseReplacer, SolutionUpdateViewModel solutionUpdateViewModel)
+    {
+      _licenseReplacer = licenseReplacer;
+      _solutionUpdateViewModel = solutionUpdateViewModel;
+    }
+
+    public string GetCommandName ()
+    {
+      return c_commandName;
+    }
+
+    public void Execute(Solution solution)
+    {
+      if (solution == null) return;
+      
+      IDictionary<string, string[]> solutionHeaderDefinitions = LicenseHeaderFinder.GetHeaderDefinitionForSolution(solution);
+
+      var allSolutionProjectsSearcher = new AllSolutionProjectsSearcher();
+      var projectsInSolution = allSolutionProjectsSearcher.GetAllProjects(solution);
+
+      var projectsWithoutLicenseHeaderFile = CheckForLicenseHeaderFileInProjects(projectsInSolution);
+
+      if (solutionHeaderDefinitions != null || !projectsWithoutLicenseHeaderFile.Any())
+      {
+        // Every project is covered either by a solution or project level license header defintion, go ahead and add them.
+        AddLicenseHeaderToProjects(projectsInSolution);
+      }
+      else
+      {
+        // Some projects are not covered by a header. Ask the user if they want to add a solution level header definition.
+        bool someProjectsHaveDefinition = projectsWithoutLicenseHeaderFile.Count != projectsInSolution.Count;
+        string question = someProjectsHaveDefinition
+          ? Resources.Question_SomeProjectsMissingAddNewLicenseHeaderDefinitionForSolution
+          : Resources.Question_AddNewLicenseHeaderDefinitionForSolution;
+
+        if (MessageBoxHelper.DoYouWant(question))
```

Comment:
It seems like you changed the behaviour quite a lot here. We would like to add the new solution header file as additional/optional feature, which does not completely replace the existing "Definition File per Project" Pattern. We can discuss where the new solution header would make sense.
 
1. If only one Project exists and it has no Definition file -> Offer to add to project (should not be changed, if we have only one project we dont need a solution wide definition file)
    1. If the offer is declined, nothing happens
    2. If the offer is approved, add file to project and offer to stop update process for configuring
        1. If this offer is declined, update license headers in project
2. If multiple Projects exists and none of them have a Definition file -> Offer to add new ones for every one of them (here it would maybe make sense to offer the solution wide License header file? I will check if my colleague has concerns regarding that change and come back to you)
    1. If the offer is declined, nothing happens
    2. If the offer is approved, add file to project and offer to stop update process for configuring
        1. If this offer is declined, update license headers in projects
3. If multiple Projects exists and some of them have a Definition file -> Offer to add existing ones for the missing projects (should not be changed, if some of them are missing a definition file it either wants to skip these projects for updating or currently uses the "definition file per project" pattern. If they want to change to solution wide, they have to do that manually.)
    1. If the offer is declined, **update the license headers for all projects**
    2. If the offer is approved, add existing file to project and offer to stop update process for configuring
        1. If this offer is declined, update license headers in all projects

Note: It seems like we currently only check for the definition file in the root of the project and dont discover them if they are inside a folder in a project. Not sure if that is a bug or a feature. This means that, to keep the current update behaviour in tact, 3. i) should definitely keep updating all projects in the solution, independent if we discovered a definition file there.

If we define the current behaviour (which i described in the "Note:") as a bug, we will fix that in a following branch. So for this pull request, keep the current behaviour, except maybe change "2." to the new solution definition file. I should have an answer from my colleague in the next hour or so.

Reply:
It sounds like we are in agreement for scenarios 1 and 2.

For point 3, my thought is that users in most cases are going to want to have the same header definitions across their solution, with specific exceptions for specific projects and folders. At least with the dozens of solutions I've seen I've never had somebody want to have different headers for different projects. It's great that it's supported, but I don't think it's the common use case.

It seems to me that the "add existing header file definition" is a feature that was designed to make sharing the same header definition file across the solution easier. But it's really not the ideal way to do it because you need to manually add the link to new projects and you need to keep the link up to date as you move projects around. It doesn't make sense to me to keep code around to continue propping up and promoting this stopgap measure when we have something better around. Adding the solution level definition is the least invasive (as you only ever need to add one file) and it future proofs you for additional projects. Prompting in this manner also might make users aware of the better alternative.

Anyway I've made the renames requested. I think the new workflow for handling missing header definitions is a lot cleaner and pushes users toward the best way of solving their problem, but let me know if you'd really want to preserve the "add existing header" workflow.

Reply:
While i agree with the general sentiment, i dont think that adding the solution definition file in this dialog is a clean solution. Would we add that file, the user would now use two different systems in his solution: the solution definition file and an unknow amount of other definition files. This could lead to confusion if he only maintains one of these systems and forgets to change the other one. 
I think that step should be actively taken and not happen because the user clicked "yes" in a dialog. Especially because the user should probably clean up the old system afterwards.

To make it clean we could offer some kind of migration scenario and offer to consilidate the existing definition files into a single solution definition file. This would have to be limited to a small scope, as it could get complex to have the same behaviour for all possible use cases. We could think about this in a follow up task.

Also, while the new solution definition file is much better for people having simple license requirements, the current system is more flexible for multiple use cases. We shouldnt make the usage harder for people who maybe need that kind of complexity.

We could inform the user of the new possibility in the dialog though. Or maybe offer both solutions in the dialog with customized buttons (altough i am not sure if that is possible with those dialogs).

Reply:
I've changed to try to prompt for an existing file when some projects have it, but others don't. But note that AddExistingLicenseHeaderDefinitionFileToProjectCommand.OpenFileDialogForExistingFile() does not seem to work for VS2017 (and is broken in your current release). That will be a separate bug for you to fix. I also added a check to filter out the "miscellaneous items" project when determining if any projects had a definition file because that showed up in the list even when there was nothing on the solution.